src/coord/rmr/cluster_topology.c

```diff
@@ -16,15 +16,36 @@ void MRClusterShard_AddNode(MRClusterShard *sh, MRClusterNode *n) {
     sh->capNodes += 1;
     sh->nodes = rm_realloc(sh->nodes, sh->capNodes * sizeof(MRClusterNode));
   }
-  sh->nodes[sh->numNodes++] = *n;
+  sh->nodes[sh->numNodes] = *n;
+  if (n->flags & MRNode_Master) {
+    // Ensure that master nodes are always at the front of the list
+    if (sh->numNodes) {
+      MRClusterNode tmp = sh->nodes[sh->numNodes];
+      sh->nodes[sh->numNodes] = sh->nodes[0];
+      sh->nodes[0] = tmp;
+    }
+  }
+  sh->numNodes++;
 }
 
-MRClusterShard MR_NewClusterShard(mr_slot_t startSlot, mr_slot_t endSlot, size_t capNodes) {
+
+void MRClusterShard_AddRange(MRClusterShard *sh, mr_slot_t start, mr_slot_t end) {
+  if (sh->capRanges == sh->numRanges) {
+    sh->capRanges++;
+    sh->ranges = rm_realloc(sh->ranges, sh->capRanges * sizeof(mr_slot_range_t));
+  }
+  sh->ranges[sh->numRanges].start = start;
+  sh->ranges[sh->numRanges].end = end;
+  sh->numRanges++;
+}
+
+MRClusterShard MR_NewClusterShard(uint32_t capNodes) {
```

Comment:
let's add capRanges here also. otherwise we are almost guaranteed to reallocate after